### DISCUSSION TITLE
Why does installing a rust tool complain that I have Python pinned???

### DISCUSSION THREAD

**Discussion by halloleo** [Homebrew/discussions#1746] [Everyday usage]
I don't understand a error message, `brew install` gives me. I have Python pinned at `python@3.9` and want to install `fd`, a tool built in Rust. So I issue:

    brew install fd

but get the error message "_Error: You must `brew unpin python@3.9` as installing fd requires the latest version of pinned dependencies_".

Why's that?  Looking at the dependcies of `fd` (via `brew info fd`) I can see me that `python` is _not_ a dependency of  `fd`.

So I don't understand why I cannot install `fd`. How can I install the rust tool and keep Python pinned at 3.9?

Thanks for any pointers!

***

### More details:

`brew install --debug --verbose fd` gives

```
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/fd.rb
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/rust.rb
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/cmake.rb
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/sphinx-doc.rb
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/python@3.9.rb
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/pkg-config.rb
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/gdbm.rb
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/mpdecimal.rb
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/openssl@1.1.rb
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/readline.rb
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/sqlite.rb
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/xz.rb
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/ninja.rb
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/libssh2.rb
Error: You must `brew unpin python@3.9` as installing fd requires the latest version of pinned dependencies
```

`brew config` gives

```
HOMEBREW_VERSION: 3.2.0
ORIGIN: https://github.com/Homebrew/brew
HEAD: 09f7bc27a99469cf947431df4754737dfbadb31d
Last commit: 7 days ago
Core tap ORIGIN: https://github.com/Homebrew/homebrew-core
Core tap HEAD: b290858939fb7ac0ebdb5ac21f51ed741a1dcc74
Core tap last commit: 5 minutes ago
Core tap branch: master
HOMEBREW_PREFIX: /usr/local
HOMEBREW_CASK_OPTS: ["--appdir=/Applications/Homebrew/"]
HOMEBREW_DISPLAY: Mac
HOMEBREW_MAKE_JOBS: 8
Homebrew Ruby: 2.6.3 => /usr/local/Homebrew/Library/Homebrew/vendor/portable-ruby/2.6.3_2/bin/ruby
CPU: octa-core 64-bit ivybridge
Clang: 11.0.0 build 1100
Git: 2.28.0 => /usr/local/bin/git
Curl: 7.54.0 => /usr/bin/curl
macOS: 10.14.6-x86_64
CLT: 10.3.0.0.1.1562985497
Xcode: 11.3.1
```

And `brew doctor` shows

```
Warning: Some installed formulae are deprecated or disabled.
You should find replacements for the following formulae:
  ilmbase

Warning: You have unlinked kegs in your Cellar.
Leaving kegs unlinked can lead to build-trouble and cause formulae that depend on
those kegs to fail to run properly once built. Run `brew link` on these:
  python@3.9
```

**Comment by carlocab** [comment]
`python@3.9` is a build dependency of `rust` (which is a build dependency of `fd`). This is likely why `brew` thinks you need the latest version of `python@3.9`.

What is the output of `brew install --debug --verbose fd`? Also, please post the output of `brew config` and `brew doctor`.

A workaround here would be to checkout the commit corresponding to your pinned version of `python@3.9` in your local homebrew/core repository and then do
```
HOMEBREW_NO_AUTO_UPDATE=1 brew install fd
```
Don't forget to checkout the `master` branch after doing that.

This sounds like a bug, though, and we should try to fix it.

**Reply by halloleo** [reply]
Thanks for chiming in, @carlocab!

I added all teh outputs to teh original question.

Regarding the workaround: Do I understand it correctly, I should checkout the corresponding commit _of the fd formula_? - Sorry to ask, but how do I do this?

**Reply by carlocab** [reply]
> I added all teh outputs to teh original question.

Sorry, but I don't see the output of `brew install --debug --verbose fd`, `brew config` or `brew doctor` in your edit.

> I should checkout the corresponding commit _of the fd formula_? 

No, you should checkout the commit corresponding to the version of `python@3.9` that you have pinned.

> how do I do this?

You need to find the bottle commit of the version of `python@3.9` you have pinned and `cd` to your local homebrew/core repo (you can find this at `$(brew --repo homebrew/core)`) and then do `git checkout <commit>`. `brew log python@3.9` might help you find the right one. You can also view the history of the `python@3.9` formula in the GitHub UI.

**Reply by halloleo** [reply]
Ups, sorry @carlocab - I had the Update only in preview. Submitted now.

Thanks for the details. Will try this out - or take Python out of brew and install "my" version of Python outside of brew. (The pinning caused me quite a bit trouble before...)

**Reply by carlocab** [reply]
`brew pin` generally causes a lot of trouble; I suggest against using it. However, the problem you have looks like a genuine bug. Fixing it isn't simple, though. Feel free to file an issue at https://github.com/Homebrew/brew so that this can be fixed.

Can I ask why you have pinned `python@3.9`? Maybe there's another solution here that we haven't discussed yet.

**Reply by halloleo** [reply]
> Can I ask why you have pinned python@3.9? Maybe there's another solution here that we haven't discussed yet.

Sure. I always had trouble when Brew updated to a new Python version that all my pipx-installed tools break: `pipx` creates virtual envs for each (Python) tool it installs, and these venvs point via symlink back to the Python binary. - I guess the symlink links to the _versioned_ binary and that always got missing through an upgrade. So then I had to re-installed every pipx-installed tool again. :-( In order to avoid this I started pinning the Python version in Brew.
